kraken/node_actions/bm_node_scenarios.py

```diff
@@ -0,0 +1,144 @@
+from kraken.node_actions.abstract_node_scenarios import abstract_node_scenarios
+import logging
+import openshift as oc
+import pyipmi
+import pyipmi.interfaces
+import sys
+import time
+import traceback
+
+class BM:
+    def __init__(self, user, passwd):
+        self.user = user
+        self.passwd = passwd
+
+    def get_node_object(self, node_name):
+        with oc.project('openshift-machine-api'):
+            return oc.selector('node/' + node_name).object()
+
+
+    # Get the ipmi or other BMC address of the baremetal node
+    def get_bmc_addr(self, node_name):
+        with oc.project('openshift-machine-api'):
+            logging.info("Getting node with name: %s" % (node_name))
+            node = self.get_node_object(node_name)
+            provider_id = node.model.spec.providerID
+            startOfUid = provider_id.rfind('/') # The / before the uid
+            startOfName = provider_id.rfind('/', 0, startOfUid) + 1
+            bmh_name = provider_id[startOfName:startOfUid]
+            bmh_resource_name = 'baremetalhost.metal3.io/' + bmh_name
+            bmh_object = oc.selector(bmh_resource_name).object()
+            return bmh_object.model.spec.bmc.address
+
+    def get_ipmi_connection(self, bmc_addr):
+        type_position = bmc_addr.find("://")
+        if type_position == -1:
+            host = bmc_addr
+        else:
+            host = bmc_addr[type_position + 3:]
+        port_position = host.find(":")
+        if port_position == -1:
+            port = 623
+        else:
+            port = int(host[port_position + 1:])
+            host = host[0:port_position]
+
+        #establish connection
+        interface = pyipmi.interfaces.create_interface('ipmitool', interface_type='lan')
+
+        connection = pyipmi.create_connection(interface)
+
+        connection.target = pyipmi.Target(ipmb_address=0x20)
+        connection.session.set_session_type_rmcp(host, port=623)
+        connection.session.set_auth_type_user(self.user, self.passwd)
+        connection.session.establish()
+        return connection
+
+
+    # Start the node instance
+    def start_instances(self, bmc_addr):
+        self.get_ipmi_connection(bmc_addr).chassis_control_power_up()
+
+    # Stop the node instance
+    def stop_instances(self, bmc_addr):
+        self.get_ipmi_connection(bmc_addr).chassis_control_power_down()
+
+    # Reboot the node instance
+    def reboot_instances(self, bmc_addr):
+        self.get_ipmi_connection(bmc_addr).chassis_control_power_cycle()
+
+    # Wait until the node instance is running
+    def wait_until_running(self, bmc_addr):
+        while (self.get_ipmi_connection(bmc_addr).get_chassis_status().power_on == False):
+            time.sleep(1)
+
+    # Wait until the node instance is stopped
+    def wait_until_stopped(self, bmc_addr):
+        while (self.get_ipmi_connection(bmc_addr).get_chassis_status().power_on == True):
+            time.sleep(1)
+
+
+class bm_node_scenarios(abstract_node_scenarios):
```

Comment:
Might want to have a method that checks connection details with given credentials before any action.